spd/clustering/scripts/run_pipeline.py

```diff
@@ -0,0 +1,384 @@
+"""Submit clustering runs to SLURM as separate jobs in a SLURM array.
+
+This script submits independent clustering runs as a SLURM job array,
+where each run gets its own dataset (seeded), WandB run, and merge history output.
+
+Also submits a job to calculate distances between the clustering runs, which will run after
+the clustering runs (the SLURM job depends on the previous array job).
+
+Output structure (only pipeline_config.json is saved to directly in this script. The files under
+<runs> are saved by run_clustering.py which is called in SLURM jobs deployed by this script.):
+    <ExecutionStamp.out_dir>/                 # from execution stamp
+        |── pipeline_config.json              # Saved in this script
+        |── clustering_run_config.json        # make copy of the file pointed to by pipeline config
+        ├── ensemble_meta.json                # (Saved by calc_distances.py) Ensemble metadata
+        ├── ensemble_merge_array.npz          # (Saved by calc_distances.py) Normalized merge array
+        ├── distances_<distances_method>.npz  # (Saved by calc_distances.py) Distance array for each method
+        └── distances_<distances_method>.png  # (Saved by calc_distances.py) Distance distribution plot
+"""
+
+import argparse
+import os
+import shlex
+import subprocess
+import tempfile
+from pathlib import Path
+from typing import Any
+
+import wandb_workspaces.workspaces as ws
+from pydantic import Field, PositiveInt
+
+from spd.base_config import BaseConfig
+from spd.clustering.consts import DistancesMethod
+from spd.clustering.storage import StorageBase
+from spd.log import logger
+from spd.utils.command_utils import run_script_array_local
+from spd.utils.general_utils import replace_pydantic_model
+from spd.utils.run_utils import _NO_ARG_PARSSED_SENTINEL, ExecutionStamp, read_noneable_str
+from spd.utils.slurm_utils import (
+    create_slurm_array_script,
+    create_slurm_script,
+    submit_slurm_script,
+)
+
+os.environ["WANDB_QUIET"] = "true"
+
+
+class ClusteringPipelineStorage(StorageBase):
+    """Storage paths for clustering pipeline (ensemble).
+
+    All paths are relative to ExecutionStamp.out_dir.
+    """
+
+    # Relative path constants
+    _PIPELINE_CONFIG = "pipeline_config.yaml"
+    _RUN_IDS = "run_ids.json"
+    _ENSEMBLE_META = "ensemble_meta.json"
+    _ENSEMBLE_MERGE_ARRAY = "ensemble_merge_array.npz"
+
+    @property
+    def pipeline_config_path(self) -> Path:
+        return self.base_dir / self._PIPELINE_CONFIG
+
+    @property
+    def run_ids_path(self) -> Path:
+        return self.base_dir / self._RUN_IDS
+
+    @property
+    def ensemble_meta_path(self) -> Path:
+        return self.base_dir / self._ENSEMBLE_META
+
+    @property
+    def ensemble_merge_array_path(self) -> Path:
+        return self.base_dir / self._ENSEMBLE_MERGE_ARRAY
+
+    def distances_path(self, method: DistancesMethod) -> Path:
+        return self.base_dir / f"distances_{method}.npz"
```

Comment:
I think @properties don't make much sense, because we don't anything to change during execution. So it gives a bit of a false impression to the reader. I think we can just define things as instance variables. This is also simpler IMO (and less code).

I made a PR for this at https://github.com/goodfire-ai/spd/pull/224. Feel free to merge.